## Classification Exercises

### Introduction

In this exercise we will learn the architecture of a neural network, the parameters of a neural network and how they change with changing architecture and the hyperparameter tuning. 

We encourage you to work with other hyperparameters as well like learning rate, number of layers, activation functions etc. 

The Notebook is divided in three parts: Building the Model, Reading the dataset and Hyperparameters. It contains five sesctions in total and one additional optional exercise:

### Part 1: Building the model
Below we define a function to built a neural network model using TensorFlow Keras. 

In [ ]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

def built_model(input_shape, n_hidden, nb_classes, optimizer='SGD'):
  '''
  input_shape: The number of inputs to the neural network
  n_hidden: Number of hidden neurons in the hidden layers
  nb_classes: Number of neurons in the output layer
  optimizer: The optimizer used to train the model. 
  By default we use Stochastic Gradient Descent.
  
  Returns:
  The function returns A model with loss and optimizer defined
  '''  
  model = tf.keras.models.Sequential()
  ## First Hidden layer  
  model.add(keras.layers.Dense(n_hidden,
       input_shape=(input_shape,),
       name='dense_layer', activation='relu'))
    
  ## Second Hidden Layer
  model.add(keras.layers.Dense(n_hidden,
        name='dense_layer_2', activation='relu'))
    
  ## Output Layer  
  model.add(keras.layers.Dense(nb_classes,
        name='dense_layer_3', activation='softmax'))
    
  ## Define loss and optimizer 
  model.compile(optimizer=optimizer, 
              loss='categorical_crossentropy',
              metrics=['accuracy'])
  return model


In [ ]:
# Defining the structure of our neural network
INPUT_SHAPE = 5
N_HIDDEN = 10
NB_CLASSES = 2  


In [ ]:
model = built_model(INPUT_SHAPE, N_HIDDEN,NB_CLASSES)

print(model.summary())

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_layer (Dense)          (None, 10)                60        
_________________________________________________________________
dense_layer_2 (Dense)        (None, 10)                110       
_________________________________________________________________
dense_layer_3 (Dense)        (None, 2)                 22        
Total params: 192
Trainable params: 192
Non-trainable params: 0
_________________________________________________________________
None


<a id='ex_2'></a>
**Here, it is important to note that our neural network has 192 parameters. As the number of networks increases, so does the complexity and execution time. Now, let's visualize the summary of the model created**

In [ ]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_layer (Dense)          (None, 10)                60        
_________________________________________________________________
dense_layer_2 (Dense)        (None, 10)                110       
_________________________________________________________________
dense_layer_3 (Dense)        (None, 2)                 22        
Total params: 192
Trainable params: 192
Non-trainable params: 0
_________________________________________________________________


### Part 2: Reading the dataset

We will continue with the MNIST dataset. 

###### Just run the cells in this part of the notebook. Do not change anything.

In [ ]:
mnist = keras.datasets.mnist
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


In [ ]:
# Processing the data
assert(len(X_train.shape)==3), "The input data is not of the right shape"
RESHAPED = X_train.shape[1]*X_train.shape[2]

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

In [ ]:
# Data Normalization
X_train, X_test = X_train / 255.0, X_test / 255.0
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

60000 train samples
10000 test samples


For the MNIST dataset the number of input and number of output units are fixed. However we can choose different values of hidden units. 

In [ ]:
INPUT_SHAPE = RESHAPED
NB_CLASSES = len(set(Y_train))

In [ ]:
# one-hot encode
Y_train = tf.keras.utils.to_categorical(Y_train, NB_CLASSES) 
Y_test = tf.keras.utils.to_categorical(Y_test, NB_CLASSES)

### Part 3: Hyperparameters

<a id='ex_3'></a>
The aim of this section is to understand the affect of changing number of hidden units on the model performance. Change the number of hidden units, and train the model. Compare the model performance in terms of accuracy. What do you understand from this?

In [ ]:
# Task to do choose different values for number of hidden units (minimum five different values)
N_HIDDEN = 10 #Choose a value

In [ ]:
## Do not change anything below
model = built_model(INPUT_SHAPE,N_HIDDEN, NB_CLASSES)
history = model.fit(X_train, Y_train,
		batch_size=128, epochs=50,
		verbose=1, validation_split=0.2)

# Evaluate the model
test_loss, test_acc = model.evaluate(X_test, Y_test)
print('Test accuracy: {:.2f} %'.format(test_acc*100))

Epoch 1/50
375/375 [==============================] - 4s 3ms/step - loss: 2.2050 - accuracy: 0.2115 - val_loss: 1.6894 - val_accuracy: 0.5091
Epoch 2/50
375/375 [==============================] - 1s 3ms/step - loss: 1.4588 - accuracy: 0.5759 - val_loss: 0.8607 - val_accuracy: 0.8032
Epoch 3/50
375/375 [==============================] - 1s 2ms/step - loss: 0.7964 - accuracy: 0.8013 - val_loss: 0.5601 - val_accuracy: 0.8568
Epoch 4/50
375/375 [==============================] - 1s 3ms/step - loss: 0.5658 - accuracy: 0.8490 - val_loss: 0.4604 - val_accuracy: 0.8749
Epoch 5/50
375/375 [==============================] - 1s 2ms/step - loss: 0.4834 - accuracy: 0.8647 - val_loss: 0.4121 - val_accuracy: 0.8872
Epoch 6/50
375/375 [==============================] - 1s 2ms/step - loss: 0.4221 - accuracy: 0.8818 - val_loss: 0.3861 - val_accuracy: 0.8913
Epoch 7/50
375/375 [==============================] - 1s 2ms/step - loss: 0.4042 - accuracy: 0.8859 - val_loss: 0.3672 - val_accuracy: 0.8959
Epoch 

In [ ]:
N_HIDDEN = [10,20,30,40,50]

In [ ]:
## Do not change anything below
for i in range(0,len(N_HIDDEN)):
  model = built_model(INPUT_SHAPE,N_HIDDEN[i], NB_CLASSES)
  history = model.fit(X_train, Y_train,
		  batch_size=128, epochs=50,
		  verbose=0, validation_split=0.2)
  # Evaluate the model
  test_loss, test_acc = model.evaluate(X_test, Y_test)
  print(test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.2995 - accuracy: 0.9195
0.9194999933242798
313/313 [==============================] - 1s 2ms/step - loss: 0.1899 - accuracy: 0.9430
0.9430000185966492
313/313 [==============================] - 1s 2ms/step - loss: 0.1650 - accuracy: 0.9523
0.9523000121116638
313/313 [==============================] - 1s 2ms/step - loss: 0.1452 - accuracy: 0.9563
0.9563000202178955
313/313 [==============================] - 1s 2ms/step - loss: 0.1478 - accuracy: 0.9548
0.954800009727478


<a id='ex_4'></a>
### Part 4: Hyperparameters

Let us now repeat the same after changing the batch size (minimum 5 different values). Compare the model performance in terms of accuracy. What do you understand from this?

**Answer** 

10: 91.94%
20: 94.30%
30: 95.23%
40: 95.63%
50: 91.94%

The results show that as we increase the number of neurons, we will obtain better results. However, it does not mean that we have increased the number of neuons in the hidden layers abruptly as it could cause an overfitting.


In [ ]:
# Task to do choose different values for batch size (minimum five different values)
BATCH_SIZE = [8,16,32,64,128] # Choose a value

In [ ]:
## Do not change anything below
model = built_model(INPUT_SHAPE,128, NB_CLASSES)
history = model.fit(X_train, Y_train,
		batch_size=BATCH_SIZE, epochs=50,
		verbose=1, validation_split=0.2)
# Evaluate the model
test_loss, test_acc = model.evaluate(X_test, Y_test)
print('Test accuracy: {:.2f} %'.format(test_acc*100))

In [ ]:
for i in range(0,len(BATCH_SIZE)):
  model = built_model(INPUT_SHAPE,128, NB_CLASSES)
  history = model.fit(X_train, Y_train,
		  batch_size=BATCH_SIZE[i], epochs=50,
		  verbose=0, validation_split=0.2)
  # Evaluate the model
  test_loss, test_acc = model.evaluate(X_test, Y_test)
  print(test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.0910 - accuracy: 0.9791
0.9790999889373779
313/313 [==============================] - 1s 2ms/step - loss: 0.0799 - accuracy: 0.9788
0.9787999987602234
313/313 [==============================] - 1s 2ms/step - loss: 0.0839 - accuracy: 0.9738
0.973800003528595
313/313 [==============================] - 1s 2ms/step - loss: 0.0890 - accuracy: 0.9728
0.9728000164031982
313/313 [==============================] - 1s 2ms/step - loss: 0.1195 - accuracy: 0.9662
0.9661999940872192


There were no big differences using different values ​​in the BATCH_SIZE parameter. As this parameter increased, the accuracy decreased slightly.

<a id='ex_5'></a>
### Part 5: Hyperparameters

**Exercise 5:** And now we do the same with different [optimizers](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers) available in TensorFlow. Change the optimizers and compare the model performance in terms of accuracy. What do you understand from this?

In [ ]:
# Task to do choose different optimizers
opt = ['sgd','adam']# Choose from available optimizers

In [ ]:
## Do not change anything below
for i in range(0,len(opt)):
  model = built_model(INPUT_SHAPE,128, NB_CLASSES, opt[i])
  history = model.fit(X_train, Y_train,
		  batch_size=128, epochs=50,
		  verbose=0, validation_split=0.2)
  # Evaluate the model
  test_loss, test_acc = model.evaluate(X_test, Y_test)
  print(test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.1239 - accuracy: 0.9635
0.9635000228881836
313/313 [==============================] - 1s 2ms/step - loss: 0.1264 - accuracy: 0.9808
0.9807999730110168


<a id='ex_O'></a>
### Optional Exercise: Fashion MNIST

Let's use a different input. You can use Fashion MNIST another popular ML dataset. Are the results same?.

To download fashion mnist you can use the following code:

```
fashion_mnist = keras.datasets.fashion_mnist

(X_train, Y_train), (X_test, Y_test) = fashion_mnist.load_data()
```

In [ ]:
fashion_mnist = keras.datasets.fashion_mnist
 
(X_train, Y_train), (X_test, Y_test) = fashion_mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


In [ ]:
# Processing the data
assert(len(X_train.shape)==3), "The input data is not of the right shape"
RESHAPED = X_train.shape[1]*X_train.shape[2]

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

In [ ]:
# Data Normalization
X_train, X_test = X_train / 255.0, X_test / 255.0
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

60000 train samples
10000 test samples


In [ ]:
INPUT_SHAPE = RESHAPED
NB_CLASSES = len(set(Y_train))

In [ ]:
# one-hot encode
Y_train = tf.keras.utils.to_categorical(Y_train, NB_CLASSES)
Y_test = tf.keras.utils.to_categorical(Y_test, NB_CLASSES)

In [ ]:
N_HIDDEN = [10,20,30,40,50]

In [ ]:
## Do not change anything below
for i in range(0,len(N_HIDDEN)):
  model = built_model(INPUT_SHAPE,N_HIDDEN[i], NB_CLASSES)
  history = model.fit(X_train, Y_train,
		  batch_size=128, epochs=50,
		  verbose=0, validation_split=0.2)
  # Evaluate the model
  test_loss, test_acc = model.evaluate(X_test, Y_test)
  print(test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.4488 - accuracy: 0.8429
0.8428999781608582
313/313 [==============================] - 1s 2ms/step - loss: 0.4230 - accuracy: 0.8515
0.8514999747276306
313/313 [==============================] - 1s 2ms/step - loss: 0.4228 - accuracy: 0.8521
0.8521000146865845
313/313 [==============================] - 1s 2ms/step - loss: 0.4029 - accuracy: 0.8564
0.8564000129699707
313/313 [==============================] - 1s 2ms/step - loss: 0.4003 - accuracy: 0.8610
0.8610000014305115


When comparing different numbers of neurons for this new data set, it is seen that the accuracy decreased compared to the results obtained with the previous data set.